Internal links: use a Monaco link provider

codebykat opened the pull request:
This PR adds a custom `LinkProvider` that searches the note for `simplenote://` links and linkifies them, then handles the click event by opening the linked note in the editor.

If the note ID does not refer to a note we can open, it will silently fail. I'm not sure what else it can or should do in that case.

Also adds a check to the preview component to skip links that can't be opened (i.e. the noteId isn't in our notes list), as otherwise it closes the selected note in favor of.. nothing.

<img width="606" alt="Screen Shot 2020-09-30 at 12 46 06 AM" src="https://user-images.githubusercontent.com/52152/94667521-c613d980-02c3-11eb-9f0e-836865471bc4.png">

codebykat committed:
use a link provider to linkify internal links

codebykat committed:
document args and limit regex to valid noteID characters

codebykat committed:
figured out how to skip the warning

belcherj commented:
Could you update the PR description with more detail for posterity? Great work!

codebykat committed:
remove types, stop overloading iLink with an additional attribute

codebykat committed:
more than zero

dmsnell commented on lib/note-content-editor.tsx:
```diff
@@ -425,6 +426,44 @@ class NoteContentEditor extends Component<Props> {
+        });
+        return {
+          links: links,
+        };
```
in some performance-critical code we can use a `for` loop to speed up iteration in JS but `.forEach` runs comparable to `.map` and in this case we're giving up semantic clarify by looping and mutating. how about a one-step transform?

```ts
return {
	// don't set URL here because then Monaco skips resolveLink
	// @cite: https://github.com/Microsoft/vscode/blob/8f89095aa6097f6e0014f2d459ef37820983ae55/src/vs/editor/contrib/links/getLinks.ts#L43:L65
	links: matches.map( ( { range } ) => ( { range } ) )
}
```

dmsnell commented on lib/note-content-editor.tsx:
```diff
@@ -425,6 +426,44 @@ class NoteContentEditor extends Component<Props> {
+
+        const [fullMatch, linkedNoteId] = match;
+        this.props.openNote(linkedNoteId as T.EntityId);
+        link.url = '#'; // tell Monaco to do nothing and not complain about it
```
let's avoid mutating the object passed-in to this function…

```ts
this.props.openNote(…);

return { ...link, url: '#' };
```

that creates a new copy instead of the "spooky action at a distance" involved when we change a property of the object passed in as an argument

codebykat committed:
Dennis is awesome

codebykat committed:
handle bad note ID

codebykat committed:
Update note-content-editor.tsx

clarify a comment

dmsnell commented on lib/note-content-editor.tsx:
```diff
@@ -425,6 +427,42 @@ class NoteContentEditor extends Component<Props> {
+        const href = editor.getModel()?.getValueInRange(link.range);
+        if (!href) {
+          return;
+        }
```
here's a stylistic nit but using the newer optional operators I find it nice to collapse multiple checks into one.

```ts
const href = editor.getModel()?.getValueInRange(link.range) ?? '';
const match = /…/.exec(href);
if (!match) {
	return;
}
```

that is, we don't care so much in this function if the `href` is empty so much as we care if it positively matches the internal link format. we can remove the special-casing for the empty link and preserve the same logic as long as we fallback to the empty string.

as I write and test that out though I realize we don't even need this. `/pattern/.exec(undefined)` and `/pattern/.exec(null)` all return `null` so we don't even  need the `?? ''` to make this focus on positively identifying matches over special-casing non-matches

codebykat replied:
I think I only did this because TypeScript was warning me about it

codebykat replied:
I don't think there's any way for this function to be called with an empty match

dmsnell replied:
> I don't think there's any way for this function to be called with an empty match

I believe that but we don't carry that information in the type system unfortunately, and hey, it's JavaSript so it could happen. we could assert that it matches but why not handle failures anyway.

`?? ''` should make TypeScript happy as well as clarify that we're matching against a string right?

dmsnell commented on lib/note-content-editor.tsx:
```diff
@@ -425,6 +427,42 @@ class NoteContentEditor extends Component<Props> {
+        const [fullMatch, linkedNoteId] = match;
+
+        if (this.props.notes.has(linkedNoteId as T.EntityId)) {
+          this.props.openNote(linkedNoteId as T.EntityId);
```
do we need to do this here or would it make sense to run it by trapping other links? I don't know the answer, I'm just asking to make sure we're not coupling the linking behavior inside the editor with app behaviors outside of the editor. for instance, do we already trap clicked-on links from the editor to intercept them?

codebykat replied:
No, since the simplenote:// links aren't real links I don't think we can trigger the link interceptor from here.

codebykat replied:
That happens over here from the preview component, but this works because they've been transformed into `<a>` tags for preview: https://github.com/Automattic/simplenote-electron/blob/b325afcf2685286d752195a862f6cad4f12ed908/lib/components/note-preview/index.tsx#L84

codebykat replied:
I guess we probably do need that check within the preview component as well, though

codebykat committed:
comment workshopping

codebykat committed:
preview component should also check for unopenable note

codebykat committed:
clean up conditional

codebykat commented:
Updated to address comments, and added a check in the preview component so that it also skips any links that can't be opened in the editor (rather than closing the selected note).

belcherj reviewed (approved):
Tested, code looks good.

dmsnell commented on lib/components/note-preview/index.tsx:
```diff
@@ -89,7 +91,12 @@ export const NotePreview: FunctionComponent<Props> = ({
             }
 
             const [fullMatch, linkedNoteId] = match;
```
if we type-cast here we should be able to avoid it redundantly below

```ts
const [fullMatch, linkedNoteId] = match as [string, T.EntityId];
```

dmsnell reviewed (approved):
Excellent work on this PR. I left a comment about making the types more convenient for you and feel free to do that and merge without a re-approval; feel free to ignore it since it's just a small convenience thing.

This was a good find allowing us to avoid the complexity of re-implementing the built-in link state machine. It's a straightforward solution.

codebykat committed:
types

souljedi commented:
Great work! I've been waiting for something like this for years.

Would it be possible to add anchor points to the link (to jump somewhere within a note)? For example the heading titles from markdown could be used?
